spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkWriteBuilder.java

```diff
@@ -116,16 +119,36 @@ public WriteBuilder overwrite(Filter[] filters) {
 
   @Override
   public Write build() {
-    // Validate
-    Schema writeSchema = validateOrMergeWriteSchema(table, dsSchema, writeConf);
+    // The write schema should only include row lineage in the output if it's an overwrite
+    // operation.
+    // In any other case, only null row IDs and sequence numbers would be produced which
+    // means the row lineage columns can be excluded from the output files
+    boolean writeIncludesRowLineage = TableUtil.supportsRowLineage(table) && overwriteFiles;
+    StructType sparkWriteSchema = dsSchema;
+    if (writeIncludesRowLineage) {
+      sparkWriteSchema = sparkWriteSchema.add("_row_id", LongType$.MODULE$);
+      sparkWriteSchema =
+          sparkWriteSchema.add(
+              MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER.name(), LongType$.MODULE$);
+    }
+
```

Comment:
Separate helper method `updateSparkWriteSchemaWithLineage` would make it clearer